File: src/main/java/io/trydent/olimpo/bus/Process.java
package io.trydent.olimpo.bus;

import io.trydent.olimpo.sys.Id;
import io.vertx.core.eventbus.EventBus;
import io.vertx.core.json.JsonObject;

import java.util.function.BiConsumer;

import static io.trydent.olimpo.vertx.json.Json.Field.field;
import static io.trydent.olimpo.vertx.json.Json.json;

public interface Process extends BiConsumer<String, Execution> {
  static Process commandProcess(Id id, EventBus bus) {
    return new CommandProcess(id, bus);
  }
}

final class CommandProcess implements Process {
  private final Id id;
  private final EventBus bus;

  CommandProcess(final Id id, final EventBus bus) {
    this.id = id;
    this.bus = bus;
  }

  @Override
  public final void accept(String name, Execution execution) {
    bus.<JsonObject>localConsumer(name, message -> {
      message.reply(
        json(
          field("id", id.get())
        ).get()
      );
      execution.accept(message.body());
    });
  }
}

File: src/main/java/io/trydent/olimpo/bus/Process.kt
package io.trydent.olimpo.bus

import io.trydent.olimpo.sys.Id
import io.trydent.olimpo.vertx.Json
import io.trydent.olimpo.vertx.json
import io.vertx.core.eventbus.EventBus

interface Process : (String, Execution) -> Unit {
  companion object {
    fun commandProcess(id: Id, bus: EventBus): Process = CommandProcess(id, bus)
  }
}

internal class CommandProcess(private val id: Id, private val bus: EventBus) : Process {
  private val message by lazy {
    json(
      "id" to id()
    )
  }

  override fun invoke(name: String, execution: Execution) {
    bus.localConsumer<Json>(name) { message ->
      message.reply(this.message)
      execution(message.body())
    }
  }
}
